framework/src/kernels/NonlocalKernel.C

```diff
@@ -0,0 +1,98 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "NonlocalKernel.h"
+#include "Assembly.h"
+#include "MooseVariable.h"
+#include "Problem.h"
+#include "SubProblem.h"
+#include "SystemBase.h"
+
+// libmesh includes
+#include "libmesh/threads.h"
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<NonlocalKernel>()
+{
+  InputParameters params = validParams<Kernel>();
+  return params;
+}
+
+NonlocalKernel::NonlocalKernel(const InputParameters & parameters) :
+    Kernel(parameters)
+{
+}
+
+NonlocalKernel::~NonlocalKernel()
+{
+}
+
+void
+NonlocalKernel::computeJacobian()
+{
+  Kernel::computeJacobian();
+
+  DenseMatrix<Number> & kenl = _assembly.jacobianBlockNonlocal(_var.number(), _var.number());
+  _nonlocal_ke.resize(kenl.m(), kenl.n());
+  _nonlocal_ke.zero();
+
+  std::set<dof_id_type> _local_dof_indices;
```

Comment:
Add comment here `// compiling set of global IDs for the local DOFs on the element`